examples/vae/net.py

```diff
@@ -1,67 +1,103 @@
+
+    def __init__(self, n_in, n_latent, n_h):
+        super(Encoder, self).__init__()
+        with self.init_scope():
+            self.linear = L.Linear(n_in, n_h)
+            self.mu = L.Linear(n_h, n_latent)
+            self.ln_sigma = L.Linear(n_h, n_latent)
+
+    def forward(self, x):
+        h = F.tanh(self.linear(x))
+        mu = self.mu(h)
+        ln_sigma = self.ln_sigma(h)  # log(sigma)
+        return D.Normal(loc=mu, log_scale=ln_sigma)
+
+
+class Decoder(chainer.Chain):
 
     def __init__(self, n_in, n_latent, n_h):
-        super(VAE, self).__init__()
+        super(Decoder, self).__init__()
         with self.init_scope():
-            # encoder
-            self.le1 = L.Linear(n_in, n_h)
-            self.le2_mu = L.Linear(n_h, n_latent)
-            self.le2_ln_var = L.Linear(n_h, n_latent)
-            # decoder
-            self.ld1 = L.Linear(n_latent, n_h)
-            self.ld2 = L.Linear(n_h, n_in)
-
-    def forward(self, x, sigmoid=True):
-        """AutoEncoder"""
-        return self.decode(self.encode(x)[0], sigmoid)
-
-    def encode(self, x):
-        h1 = F.tanh(self.le1(x))
-        mu = self.le2_mu(h1)
-        ln_var = self.le2_ln_var(h1)  # log(sigma**2)
-        return mu, ln_var
-
-    def decode(self, z, sigmoid=True):
-        h1 = F.tanh(self.ld1(z))
-        h2 = self.ld2(h1)
-        if sigmoid:
-            return F.sigmoid(h2)
-        else:
-            return h2
-
-    def get_loss_func(self, beta=1.0, k=1):
-        """Get loss function of VAE.
-
-        The loss value is equal to ELBO (Evidence Lower Bound)
-        multiplied by -1.
-
-        Args:
-            beta (int): Usually this is 1.0. Can be changed to control the
-                second term of ELBO bound, which works as regularization.
-            k (int): Number of Monte Carlo samples used in encoded vector.
-        """
-        def lf(x):
-            mu, ln_var = self.encode(x)
-            batchsize = len(mu.data)
-            # reconstruction loss
-            rec_loss = 0
-            for l in six.moves.range(k):
-                z = F.gaussian(mu, ln_var)
-                rec_loss += F.bernoulli_nll(x, self.decode(z, sigmoid=False)) \
-                    / (k * batchsize)
-            self.rec_loss = rec_loss
-            self.loss = self.rec_loss + \
-                beta * gaussian_kl_divergence(mu, ln_var) / batchsize
-            chainer.report(
-                {'rec_loss': rec_loss, 'loss': self.loss}, observer=self)
-            return self.loss
-        return lf
+            self.linear = L.Linear(n_latent, n_h)
+            self.output = L.Linear(n_h, n_in)
+
+    def forward(self, z, inference=False):
+        n_batch_axes = 1 if inference else 2
+        h = F.tanh(self.linear(z, n_batch_axes=n_batch_axes))
+        h = self.output(h, n_batch_axes=n_batch_axes)
+        return D.Bernoulli(logit=h)
+
+
+class Prior(chainer.Chain):
+
+    def __init__(self, n_latent, dtype=np.float32, device=-1):
```

Comment:
Please do not specify `device` here.  Why don't you make `Prior` inherit `chainer.Link` and add `loc` and `scale` as `chainer.Parameter`s?

Reply:
I fixed it: 39f8620

Reply:
I have just realized that the prior distribution of the original VAE paper is not trainable. My current implementation (which is based on your advice) makes `loc` and `scale` as `chainer.Parameter`s and it makes these variables trainable. I couldn't find how to allocate non-trainable variable without using `device` option. Are there any ways to do that?

Reply:
Could you try `register_persistent`?

Reply:
parameters for prior distribution now seem to be non-trainable, thank you. 8294d6f